index.js

```diff
@@ -0,0 +1,214 @@
+const roundReward = 456621004566210048n
+
+const handler = async (req, res, redis, signerAddresses) => {
+  if (req.method === 'POST' && req.url === '/scores') {
+    await handleIncreaseScores(req, res, redis, signerAddresses)
+  } else if (req.method === 'POST' && req.url === '/paid') {
+    await handlePaidScheduledRewards(req, res, redis, signerAddresses)
+  } else if (req.method === 'GET' && req.url === '/scheduled-rewards') {
+    await handleGetAllScheduledRewards(res, redis)
+  } else if (req.method === 'GET' && req.url.startsWith('/scheduled-rewards/')) {
+    await handleGetSingleScheduledRewards(req, res, redis)
+  } else if (req.method === 'GET' && req.url === '/log') {
+    await handleGetLog(res, redis)
+  } else {
+    status(res, 404)
+  }
+}
+
+const validateSignature = (signature, participants, signerAddresses) => {
+  httpAssert(
+    typeof signature === 'object' && signature !== null,
+    400,
+    '.signature should be an object'
+  )
+  httpAssert.deepEqual(
+    Object.keys(signature).sort(),
+    ['r', 's', 'v'],
+    400,
+    '.signature should have keys .r, .s and .v'
+  )
+
+  const digest = ethers.solidityPackedKeccak256(
+    ['address[]', 'int256[]'],
+    [Object.keys(participants), Object.values(participants)]
+  )
+  const reqSigner = ethers.verifyMessage(
+    digest,
+    ethers.Signature.from(signature)
+  )
+  httpAssert(signerAddresses.includes(reqSigner), 403, 'Invalid signature')
+}
+
+async function handleIncreaseScores (req, res, redis, signerAddresses) {
+  const body = JSON.parse(await getRawBody(req, { limit: '1mb' }))
+
+  httpAssert(
+    typeof body === 'object' && body !== null,
+    400,
+    'Request body should be an object'
+  )
+  httpAssert(
+    typeof body.scores === 'object' && body.scores !== null,
+    400,
+    '.scores should be an object'
+  )
+  httpAssert(
+    Object.keys(body.scores).every(ethers.isAddress),
+    400,
+    'All .scores keys should be 0x addresses'
+  )
+  httpAssert(
+    Object.values(body.scores).every(score => {
+      try {
+        return BigInt(score) > 0n
+      } catch {
+        return false
+      }
+    }),
+    400,
+    'All .scores values should be positive numbers encoded as string'
+  )
+
+  validateSignature(body.signature, body.scores, signerAddresses)
+
+  const timestamp = new Date()
+  const tx = redis.multi()
+  for (const [address, score] of Object.entries(body.scores)) {
+    const scheduledRewards = (BigInt(score) * roundReward) / maxScore
+    tx.hincrby('rewards', address, scheduledRewards)
+    tx.rpush(
+      'log',
+      JSON.stringify({
+        timestamp,
+        address,
+        score,
+        scheduledRewards: String(scheduledRewards)
+      })
+    )
+  }
+  const updated = await tx.exec()
+
+  json(
+    res,
+    Object.fromEntries(
+      Object
+        .keys(body.scores)
+        .map((address, i) => [address, String(updated[i * 2][1])])
```

Comment:
I am confused: how did you arrive at the formula `updated[i * 2][1]`? 

When I try to unpack this:
- Is `updated` an array of Redis transactions performed? Let's give it a different name then, e.g. `responses` or `results`.
- I assume that `i*2` is because we create two transactions for each participant (`hincrby` and `rpush`). Could you please add a code comment to explain?
- Then I assume that `[1]` is to find the response/result field that contains the updated value. Could you please add a code comment to list all field names in the array or to link to the Redis doc showing these fields?

Reply:
This is correct: Every 2nd entry is from `hincrby`, which returns the new value. Inside the array, there are two fields, and the 2nd contains the result. I inspected the returned value to gain this information.

I've added a comment about this 👍